overlord/snapstate/autorefresh.go

```diff
@@ -682,6 +688,116 @@ func (e *timedBusySnapError) Is(err error) bool {
+			logger.Noticef("Cannot send notification about finishing deferred refresh: %v", err)
+		}
+	}()
+}
+
+// asyncFinishDeferredRefreshNotification broadcasts desktop notification in a goroutine,
+// telling the user that a refresh process has been completed and that they can now launch
+// the snap.
+//
+// This allows the, possibly slow, communication with each snapd session agent,
+// to be performed without holding the snap state lock.
+var asyncFinishDeferredRefreshNotification = func(context context.Context, client *userclient.Client, finishInfo *userclient.FinishedSnapRefreshInfo) {
+	go func() {
+		if err := client.FinishDeferredRefreshNotification(context, finishInfo); err != nil {
+			logger.Noticef("Cannot send notification about finishing deferred refresh: %v", err)
+		}
+	}()
+}
+
+// Adds a Snap to the list of snaps to monitor when all their instances
+// are closed. When that happens, the name of that snap is sent through
+// the channel.
+var MonitorSnap = func(snapName string, monitorChannel chan string) {
+	cgroup.MonitorSnapEnded(snapName, monitorChannel)
+}
+
+// Sends the 'refresh' command for a snap after StatusChecker has detected
+// that all the instances have been closed.
+// When it ends, it sends the result of the command through workerChannel,
+// to notify the main thread that the operation has ended and that another
+// refresh can be launched. This allows to serialize several refreshes,
+// avoiding to launch all of them at the same time.
+var RefreshSnap = func(snapName string, desktopPath string) {
+	deferredRefreshInfo := userclient.FinishedSnapRefreshInfo{InstanceName: snapName, AppDesktopEntry: desktopPath}
+	clientConnection := userclient.New()
+	asyncBeginDeferredRefreshNotification(context.TODO(), clientConnection, &deferredRefreshInfo)
+	clientConfig := client.Config{
+		// we need the powerful snapd socket
+		Socket: dirs.SnapdSocket,
+	}
+	snapOpts := client.SnapOptions{}
+	cl := client.New(&clientConfig)
+	changeId, err := cl.Refresh(snapName, &snapOpts)
+
+	if err != nil {
+		asyncFinishDeferredRefreshNotification(context.TODO(), clientConnection, &deferredRefreshInfo)
+		return
+	}
+	clientRefreshStatus := userclient.SnapRefreshProgressInfo{InstanceName: snapName}
+	for true {
+		change, err := cl.Change(changeId)
+		if (err != nil) || (change.Status == "Done") {
+			asyncFinishDeferredRefreshNotification(context.TODO(), clientConnection, &deferredRefreshInfo)
+			break
+		}
+		tasksDone := 0
+		newValue := -1.0
+		newBarText := ""
+		for taskId := range change.Tasks {
+			task := change.Tasks[taskId]
+			if task.Status == "Done" {
+				tasksDone++
+			}
+			if (task.Status == "Doing") && (newBarText == "") {
+				newBarText = task.Summary
+				if task.Progress.Total > 1 {
+					newValue = float64(task.Progress.Done) / float64(task.Progress.Total)
+				}
+			}
+		}
+		if newValue == -1 {
+			newValue = float64(tasksDone) / float64(len(change.Tasks))
+		}
```

Comment:
These lines could be moved to a helper `calculateChangeProgress`, which returns `pct, barText := calculateChangeProgress(change)`

Reply:
Done.